### ISSUE TITLE
urxvt doesn't display progress bars correctly

### ISSUE THREAD

**Issue by vikstrous** [docker/docker#28111]
**Description**

When using urxvt to pull an image the progress bars are drawn off by one line. This is annoying because it causes statically printed messages to be overriden by progress bars and it can make the status of certain layers get covered up.

**Steps to reproduce the issue:**
1. docker pull an image with more than one layer using urxvt

**Describe the results you received:**

```
$ docker pull ascribe/image-match
Using default tag: latest
latest: Pulling from ascribe/image-match

43c265008fae: Already exists 
af36d2c7a148: Already exists 
f6a5aab6cd0c: Downloading 539.8 kB/129.8 MB
132c9d275259: Downloading 719.6 kB/2.889 MB
f2bfc1fc1d76: Downloading 654.6 kB/15.88 MB
5a0288588dea: Waiting 
276aa5e3c7b2: Waiting 
0a789057158f: Waiting 
b579dcc73c1a: Waiting 
9fca55327c3f: Waiting 
60402126d860: Waiting 
949bc7382b9d: Waiting 
25784ad4ced1: Waiting 
25784ad4ced1: Pulling fs layer
```

Note that the last line here is printed on the same line as the cursor when using urxvt, but it's printed one line above the cursor when using xterm. That seems to be the source of the issue.

**Describe the results you expected:**
```
$ docker pull ascribe/image-match
Using default tag: latest
latest: Pulling from ascribe/image-match

43c265008fae: Already exists 
af36d2c7a148: Already exists
143e9d501644: Already exists
f6a5aab6cd0c: Downloading 539.8 kB/129.8 MB
132c9d275259: Downloading 719.6 kB/2.889 MB
f2bfc1fc1d76: Downloading 654.6 kB/15.88 MB
5a0288588dea: Waiting 
276aa5e3c7b2: Waiting 
0a789057158f: Waiting 
b579dcc73c1a: Waiting 
9fca55327c3f: Waiting 
60402126d860: Waiting 
949bc7382b9d: Waiting 
25784ad4ced1: Waiting 
```

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
Client:
 Version:      1.12.2-cs2
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   d5fda6e
 Built:        Sun Oct 23 07:48:43 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.13.0-dev
 API version:  1.25
 Go version:   go1.7.3
 Git commit:   c5ceb0f-unsupported
 Built:        Sun Nov  6 06:37:06 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 6
 Running: 1
 Paused: 0
 Stopped: 5
Images: 82
Server Version: 1.13.0-dev
Storage Driver: overlay2
 Backing Filesystem: extfs
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: apparmor seccomp
Kernel Version: 4.8.3
Operating System: NixOS 17.03.git.210b3b3 (Gorilla)
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 15.53 GiB
Name: thisisfine
ID: 4XB6:RAOL:2UT7:W3TT:JBSR:ILHC:LREA:HUVU:5U7V:TI7E:WUT2:WQTS
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 22
 Goroutines: 34
 System Time: 2016-11-06T15:07:26.849363997-08:00
 EventsListeners: 0
Username: viktorstanchev
Registry: https://index.docker.io/v1/
WARNING: No cpu cfs quota support
WARNING: No cpu cfs period support
Insecure Registries:
 172.17.0.1
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
physical

**Comment by ripcurld00d** [comment]
@vikstrous strange. I am using urxvt(+tmux) but can't reproduce this error.

```shell
$ urxvt --help 2>&1 | head -1
rxvt-unicode (urxvt) v9.21 - released: 2014-12-31
$ docker pull ascribe/image-match
Using default tag: latest
latest: Pulling from ascribe/image-match

43c265008fae: Already exists 
af36d2c7a148: Already exists 
143e9d501644: Already exists 
f6a5aab6cd0c: Pull complete 
132c9d275259: Pull complete 
f2bfc1fc1d76: Pull complete 
5a0288588dea: Pull complete 
276aa5e3c7b2: Pull complete 
0a789057158f: Pull complete 
b579dcc73c1a: Pull complete 
9fca55327c3f: Pull complete 
60402126d860: Pull complete 
949bc7382b9d: Pull complete 
25784ad4ced1: Pull complete 
Digest: sha256:516ff848529f0959e52a16193fb3f3b6509bd28928fc388734b550c99ae0a7c3
Status: Downloaded newer image for ascribe/image-match:latest
```

**Comment by ijc25** [comment]
I see this with `urxvt` (`rxvt-unicode` package version 9.22-1 in Debian) but not if I run `tmux` and do the pull there. @ripcurld00d I'm guessing your use of `tmux` is hiding the issue from you (makes sense since I think it is involve in handling the vt100 escape codes etc)

**Comment by ripcurld00d** [comment]
@ijc25 thanks for clarifying this.

**Comment by ijc25** [comment]
Capturing some output with `script(1)` I think the problem may be with the sequence `<ESC>[1B<ESC>[1A^M`. (Where `<ESC>` is actually the ASCII ESC character 0x1b and `^M` is carriage return 0x0d)

According to [Wikipedia](https://en.wikipedia.org/wiki/ANSI_escape_code#CSI_codes) `<ESC>[1B` is "Move cursor down", with the caveat "If the cursor is already at the edge of the screen, this has no effect.". `<ESC>[1A` is likewise move up 1 unless at the edge.

It appears that the sequence `First<ESC>[1B<ESC>[1A^MSecond` on urxvt produces (where _ is the final position of the cursor)
```
Second_
First
```
Compared with xterm which instead produces
```
Second_
```
because (I'm speculating) `<ESC>[1B` does nothing (already at the edge of the screen), `<ESC>[1A` moves up one line, then `^M` moves back down and "Second" overwrites the "First".

I'm not sure what exactly is happening with urxvt to cause the observed behaviour, which is that after  `<ESC>[1B<ESC>[1A^M` the cursor has moved up one line.

I'm not even really convinced as to which of the two is right or wrong. Might be best to take this one to the urxvt list perhaps?

**Comment by ijc25** [comment]
I tested with aterm, xterm and mlterm and they all behave correctly. I'm going to post to the rxvt-unicode ML and see if they have any insights.

**Comment by ijc25** [comment]
I received a series of very helpful replies from Marc on the rxvt-unicode list:
http://lists.schmorp.de/pipermail/rxvt-unicode/2016q4/002315.html

His conclusion:
> In summary: docker outputs a lot of "move cursor 0 characters" commands, which are undocumented, non-ANSI, and behave different in different terminals.

He also recommends using termcap, which seems very sensible to me, I'll take a look at that and try to raise a PR.

**Comment by ijc25** [comment]
Ooops, `s/termcap/terminfo/` in my parts of the above.

**Comment by ijc25** [comment]
I proposed a fix for the underlying issue in #28238, I'm also still looking into using terminfo (oh man, my fingers typed term*cap* the first time *again*).

**Comment by ijc25** [comment]
#28238 has been merged and fixed the underlying issue here. I think this issues can be closed. I'm still looking at a switch to terminfo but no need to keep this open for that I think.
